{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0aadf490-986a-4e5e-85e6-a6e2cfaee5de",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<h2>Bioinformatics</h2>\n",
    "<p>Dan Vogel</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53dfa411-affe-4c3b-adfc-4ac066b90fd0",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<b>About me</b>\n",
    "<p>\n",
    "Getting Started and continuing education...<br>\n",
    "Took programming classes at Montgomery College during senior year in HS. Worked at College Computer lab. \n",
    "Studied at Montgomery College in transfer program. Earned B.S. in Computer Science at UMD in College Park. \n",
    "Worked as Student Systems Programmer then became a Systems Programmer at UMUC, upon graduation.\n",
    "Earned MBA while working at UMUC and much later, Graduate certicate, Information Assurance. In 2023, earned MS Bioinformatics at Hood College.\n",
    "\n",
    "Professional certifications include CISSP, PMP, Red Hat Certified, CCNA and many CompTIA+ certifications.\n",
    "</p>\n",
    "<p>\n",
    "Current title is Manager, High Performance Computing (HPC) at National Biodefense Analysis and Countermeasures Center (NBACC). <br>\n",
    "<a href=\"https://www.dhs.gov/science-and-technology/national-biodefense-analysis-and-countermeasures-center\" NBACC Website</a></p>\n",
    "I manage 2 linux server clusters used to process Bioinformatics data.\n",
    "</p>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b7cb42-a196-4f09-b833-63cad354b642",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<p>\n",
    "Previously:\n",
    "<ul>\n",
    "    <li>Lead Systems Administrator at US Army Medical Research institute (USAMRIID). <br>\n",
    "        <a href=\"https://usamriid.health.mil\">https://usamriid.health.mil</a>\n",
    "    </li>\n",
    "    <li>Project Manager for Cybersecurity Operations at the US Coast Guard Operation Systems Center. <br>\n",
    "        <a href=\"https://www.dcms.uscg.mil/Our-Organization/Assistant-Commandant-for-C4IT-CG-6/Operations-Systems-Center-OSC/\">USCG OSC Website</a></li>\n",
    "    <li>Director of Technology programs at Hagerstown Community college.<br>\n",
    "        <a href=\"https://www.hagerstowncc.edu/\">HCC Website</a></li>\n",
    "    <li>Director for IT at University of Maryland Asia <br><a href=\"https://asia.umgc.edu/\">UMGC Asia Website</a></li>\n",
    "</ul>\n",
    "</p>\n",
    "<p>\n",
    "Personal Links\n",
    "<ul>\n",
    "    <li><a href=\"https://www.linkedin.com/in/danieltvogel\">linkedin.com/in/danieltvogel</a></li>\n",
    "    <li><a href=\"https://github.com/japandan\">github.com/japandan</a></li>\n",
    "</ul>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "579b550f-f9fd-46c6-8e10-0adc686c96c5",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<b>Google says...</b><br>\n",
    "<p>\n",
    "\"Bioinformatics is an interdisciplinary field that merges biology, computer science, statistics, and information technology to analyze, interpret, and manage complex biological data, such as DNA, RNA, and protein sequences. It is critical for genomics, drug discovery, and molecular modeling, enabling scientists to uncover patterns and gain insights into molecular function.\"</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93db7463-35e9-48f1-8ed2-fddf1cb39014",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<b>Put another way...</b><br>\n",
    "<p>Bioinformatics is essentially BIG data analysis for biological data sets. \n",
    "It requires computational and statistical analyses in order to extract meaning from biological data. \n",
    "Since this is the case, bioinformatics can also be used to refer to the development of the software and methods which are used to understand biological data...</p>\n",
    "    <img src=\"./images/Bioinformatics-Position-400x400.png\" alt=\"Venn Diagram\"><br>\n",
    "    <a href=\"url\">https://www.fiosgenomics.com/what-is-bioinformatics-overview-and-examples/</a>                                                                   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33dd9769-2069-4b17-8eba-5a997653cab6",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# TedTalk Using Bioinformatics to cure disease\n",
    "<a href=\"url\">https://www.youtube.com/watch?v=_eHz6qzTCfc</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e19263",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## The Central Dogma of Molecular Biology:DNA->RNA-> PROTEIN\n",
    "<p>\n",
    "DNA is transcribed into RNA, which is then translated into protein. This fundamental process ensures that instructions in the genome are executed to build cellular structures and perform functions.\n",
    "<p>\n",
    "    <img src=\"./images/Central_Dogma_of_Molecular_Biochemistry_with_Enzymes.jpg\" alt=\"Dogma diagram\">\n",
    "</p>\n",
    "<p>\n",
    "    <a href=\"https://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology\">https://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology</a>\n",
    "</p>\n",
    "<p>\n",
    "    <a href=\"https://flexbooks.ck12.org/cbook/ck-12-biology-flexbook-2.0/section/4.1/primary/lesson/central-dogma-of-molecular-biology-bio/\">FlexBooks animation</a>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d98a4df-b8ef-4baf-a6a0-019869e15eda",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# DNA Alphabet form the rungs of the double helix ladder\n",
    "## A,T,G,C\n",
    "### A = adenine, T = Thymine,   \n",
    "### G = Guanine, C = Cytosine\n",
    "<pre>\n",
    "  --+--+--+--+--+--+--+--\n",
    "    A  T  A  C  A  G  T\n",
    "    +  +  +  +  +  +  +   \n",
    "    T  A  T  ?  ?  ?  ?\n",
    "  --+--+--+--+--+--+--+--\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "037cacc3-f049-478f-9008-d36cae040a4a",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# RNA Alphabet ( the T are replaced by U ) on messenger RNA\n",
    "## A,U,G,C\n",
    "<pre>\n",
    "  --+--+--+--+--+--+--\n",
    "    A  U  A  C  A  G  \n",
    "</pre>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20ad1d57-c4b5-4837-93f3-12ec36484ba4",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## Codons are 3 bases together and they match up with an amino acid to make a protein\n",
    "### Example: codon GCU = Alanine = letter A in the protein alphabet\n",
    "\n",
    "There are 2 codons in this RNA.  AUA & CAG\n",
    "<pre>\n",
    "  --+--+--+--+--+--+--\n",
    "    A  U  A  C  A  G  \n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ca7673-4877-47ed-b98e-f893df47e49a",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## Protein Alphabet has 20 letters, each is an amino acid\n",
    "\n",
    "### 'A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y' "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71dfa7cd-bc7d-4c7f-9823-50e1f5254c03",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "##  The genetic code consists of 64 triplets (codons) of nucleotides.\n",
    "### 61 coding for 20 amino acids and 3 acting as stop signals.\n",
    "### Multiple codons can specify the same amino acid. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48047cf4-2341-434a-850a-8f2f59f67219",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "- What amino acid gets created from codon AUA ?\n",
    "- What amino acid gets created from codon CAG ?\n",
    " \n",
    "Standard Amino Acid Codon List (mRNA) \n",
    "\n",
    "    Alanine (Ala, A): GCU, GCC, GCA, GCG\n",
    "    Arginine (Arg, R): CGU, CGC, CGA, CGG, AGA, AGG\n",
    "    Asparagine (Asn, N): AAU, AAC\n",
    "    Aspartic acid (Asp, D): GAU, GAC\n",
    "    Cysteine (Cys, C): UGU, UGC\n",
    "    Glutamine (Gln, Q): CAA, CAG\n",
    "    Glutamic acid (Glu, E): GAA, GAG\n",
    "    Glycine (Gly, G): GGU, GGC, GGA, GGG\n",
    "    Histidine (His, H): CAU, CAC\n",
    "    Isoleucine (Ile, I): AUU, AUC, AUA  \n",
    "    Leucine (Leu, L): UUA, UUG, CUU, CUC, CUA, CUG\n",
    "    Lysine (Lys, K): AAA, AAG\n",
    "    Methionine (Met, M) [Start]: AUG\n",
    "    Phenylalanine (Phe, F): UUU, UUC\n",
    "    Proline (Pro, P): CCU, CCC, CCA, CCG\n",
    "    Serine (Ser, S): UCU, UCC, UCA, UCG, AGU, AGC\n",
    "    Threonine (Thr, T): ACU, ACC, ACA, ACG\n",
    "    Tryptophan (Trp, W): UGG\n",
    "    Tyrosine (Tyr, Y): UAU, UAC\n",
    "    Valine (Val, V): GUU, GUC, GUA, GUG "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aff59ee-4d7e-47ad-b1d3-9c996f932f0b",
   "metadata": {
    "editable": true,
    "raw_mimetype": "slides",
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<code>\n",
    "# Example python code to generate condons or kmer\n",
    "def generate_trimers():\n",
    "    bases = ['A','U','G','C']\n",
    "    result = []\n",
    "    for base1 in bases:\n",
    "        for base2 in bases:\n",
    "            for base3 in bases:\n",
    "                result.append( base1 + base2 + base3 )\n",
    "    return result\n",
    "</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c92a89f0-0055-4311-824c-75517d44315e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAA AAU AAG AAC AUA AUU AUG AUC AGA AGU AGG AGC ACA ACU ACG ACC UAA UAU UAG UAC UUA UUU UUG UUC UGA UGU UGG UGC UCA UCU UCG UCC GAA GAU GAG GAC GUA GUU GUG GUC GGA GGU GGG GGC GCA GCU GCG GCC CAA CAU CAG CAC CUA CUU CUG CUC CGA CGU CGG CGC CCA CCU CCG CCC\n"
     ]
    }
   ],
   "source": [
    "# Example python code to generate condons or kmer\n",
    "def generate_trimers():\n",
    "    bases = ['A','U','G','C']\n",
    "    result = []\n",
    "    for base1 in bases:\n",
    "        for base2 in bases:\n",
    "            for base3 in bases:\n",
    "                result.append( base1 + base2 + base3 )\n",
    "    return result\n",
    "rna=generate_trimers()\n",
    "print(*rna)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cc63b02-3c7c-43ca-8076-6fe130ae7338",
   "metadata": {
    "editable": true,
    "raw_mimetype": "slides",
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "<code>\n",
    "# Using recursion, and create kmer of any length \n",
    "def generate_kmers_rec(length):\n",
    "    if length == 1:\n",
    "        return ['A','U','G','C']\n",
    "    else:\n",
    "        result = []\n",
    "        for seq in generate_kmers_rec( length - 1 ):\n",
    "            for base in ['A','U','G','C']:\n",
    "                result.append( seq + base )\n",
    "        return result\n",
    "</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cd14f25e-4750-4538-86a0-1ab7a936a7f9",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAA AAU AAG AAC AUA AUU AUG AUC AGA AGU AGG AGC ACA ACU ACG ACC UAA UAU UAG UAC UUA UUU UUG UUC UGA UGU UGG UGC UCA UCU UCG UCC GAA GAU GAG GAC GUA GUU GUG GUC GGA GGU GGG GGC GCA GCU GCG GCC CAA CAU CAG CAC CUA CUU CUG CUC CGA CGU CGG CGC CCA CCU CCG CCC\n"
     ]
    }
   ],
   "source": [
    "# Using recursion, and create kmer of any length \n",
    "def generate_kmers_rec(length):\n",
    "    if length == 1:\n",
    "        return ['A','U','G','C']\n",
    "    else:\n",
    "        result = []\n",
    "        for seq in generate_kmers_rec( length - 1 ):\n",
    "            for base in ['A','U','G','C']:\n",
    "                result.append( seq + base )\n",
    "        return result\n",
    "\n",
    "rna = generate_kmers_rec(3)\n",
    "print( *rna )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5becd46-6054-47a7-8138-61b4d02aaf23",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Programming resources dedicated to bioinformatics\n",
    "### Python - Bioconda channel, Conda-forge channel\n",
    "### R - Bioconductor\n",
    "### Kaggle datasets  https://www.kaggle.com/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99615249-f7ef-4646-be28-b8d3cf181653",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# AI and Protein folding : Alphafold\n",
    "AlphaFold is\n",
    "a revolutionary artificial intelligence system developed by Google DeepMind that predicts a protein's 3D structure from its 1D amino acid sequence with near-experimental accuracy. By solving the 50-year-old \"protein folding problem,\" AlphaFold has transformed structural biology, accelerating research in drug discovery, disease understanding, and biotechnology. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f08265a-a973-4d82-99fe-ca5c8491404a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Bioinformatics and COVID\n",
    "- DNA samples taken all over the world as cases were found. \n",
    "- Allowed scientists to track the spread and when new strains were found\n",
    "- Allowed scientists to develop mRNA vaccines "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10b6a2e-7d81-412e-b493-d83d35a43387",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Where are the jobs?\n",
    "- National and private labs: NIH, NCI, USAMRIID, NBACC, Los Alamos,Brookhaven\n",
    "- Pharmaceuticals: Qiagen\n",
    "- Life sciences companies: Illumina, PacBio\n",
    "- Research at Universities\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc11466-20b8-4c61-9ec8-296d4a3d2ec3",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Data Science\n",
    "* Financial companies, predicting stocks\n",
    "* Insurance companies, modeling\n",
    "* Institutional Research, Business"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b473bd4-361f-4641-ad8f-d7da3a33a072",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# What education is required?\n",
    "\n",
    "## Bioinformatics\n",
    "+ Usually a BS in Computer Science or Biology\n",
    "+ Statistics class\n",
    "+ Programming experience: Python, R\n",
    "+ MS and PHd to lead research\n",
    "+ DevOps skills\n",
    "\n",
    "## IT Jobs and WetLab\n",
    "+ High School diploma\n",
    "+ Certifications: A+, Security+, Network+\n",
    "+ Scripting skills: Powershell, Bash, Python\n",
    "+ WetLab skills: sample preparations, safety protocols"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "881914f2",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    " <img src=\"./jobs/Bioinformatics.scientist.png\" alt=\"Scientist Salaries\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7fc9198",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    " <img src=\"./jobs/JHU.job.titles.png\" alt=\"Job Titles\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b95ad794-cb95-4df9-917d-792e8ded47b8",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Questions?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
